File: python/backtracking/generate_parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result=  []


        def backtrack(current: str, open: int, close: int):
            if len(current) == 2 * n:
                result.append(current)

                return

            if open < n:
                backtrack(current + "(", open+1, close)

            if close < open:
                backtrack(current + ")", open, close+1)


        backtrack("", 0, 0)

        return result

File: golang/backtracking/generate_parentheses.go
package main

func generateParenthesis(n int) []string {
	var result []string

	var backtrack func(current string, open, close int)

	backtrack = func(current string, open, close int) {
		if len(current) == 2*n {
			result = append(result, current)

			return
		}

		if open < n {
			backtrack(current+"(", open+1, close)
		}

		if close < open {
			backtrack(current+")", open, close+1)
		}
	}

	backtrack("", 0, 0)

	return result
}
